File: 백준/Gold/2467. 용액/용액.cc
#include <iostream>
using namespace std;
int arr[100000];
int ans = 2000000001, ansl, ansr;
int main() {
	int n;
	cin >> n;
	for (int i = 0; i < n; i++) {
		cin >> arr[i];
	}
	for (int i = 0; i < n; i++) {
		int start = 0;
		int end = n - 1;
		int mid = (start + end) / 2;
		while (start <= end) {
			mid = (start + end) / 2;
			if (arr[i] + arr[mid] > 0) {
				end = mid - 1;
			}
			else if (arr[i] + arr[mid] < 0) {
				start = mid + 1;
			}
			else {
				break;
			}
		}

		for (int j = mid - 1; j <= mid + 1; j++) {
			if (j < 0 || j >= n || i == j)continue;
			if (ans > abs(arr[i] + arr[j])) {
				ans = abs(arr[i] + arr[j]);
				ansl = i;
				ansr = j;
			}
		}
	}
	if (arr[ansl] > arr[ansr])swap(ansl, ansr);
	cout << arr[ansl] << " " << arr[ansr];
}

File: 백준/Gold/2467. 용액/용액.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {

	public static void main(String[] args) throws Exception {
		var br = new BufferedReader(new InputStreamReader(System.in));

		int n = Integer.parseInt(br.readLine());
		int[] a = new int[n];
		var st = new StringTokenizer(br.readLine());
		for (int i = 0; i < n; i++) {
			a[i] = Integer.parseInt(st.nextToken());
		}

		int l = 0, r = n - 1;
		int ans = Integer.MAX_VALUE;
		int ansl = a[0], ansr = a[n - 1];
		while (l < r) {
			int abs = Math.abs(a[l] + a[r]);
			if (ans > abs) {
				ans = abs;
				ansl = a[l];
				ansr = a[r];
			}
			if (a[l] + a[r] > 0) {
				r--;
			} else {
				l++;
			}
		}
		System.out.println(ansl + " " + ansr);
	}
}